### PR TITLE
Group Cryptography: Bugs fixed

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
Group decryption fails on LWIP platforms. This problem is caused by:
* Operational keys stored incorrectly
* Invalid pointer passed to SymmetricKeyContext::DecryptMessage() in LWIP platforms.

#### Change overview
* Operational keys' storage fixed in GroupDataProviderImpl::SetKeySet
* SymmetricKeyContext interface modified to allow separate input and output buffers
* GroupDataProviderImpl updated to use the new SymmetricKeyContext interface
* CryptoContext updated to use the new  SymmetricKeyContext interface
* TestGroupDataProvider updated.

#### Testing
* Project built successfully
* TestGroupMessaging ran successfully
* TestGroupDataProvider tests ran successfully
* TestWriteInteraction tests ran successfully

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by jepenven-silabs — APPROVED** [review]

File: src/credentials/GroupDataProviderImpl.cpp [hunk 1]
```diff
@@ -1606,12 +1607,11 @@ CHIP_ERROR GroupDataProviderImpl::SetKeySet(chip::FabricIndex fabric_index, cons
     for (size_t i = 0; i < in_keyset.num_keys_used; ++i)
     {
         ByteSpan epoch_key(in_keyset.epoch_keys[i].key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
-        uint8_t key[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
-        MutableByteSpan key_span(key, sizeof(key));
+        MutableByteSpan key_span(keyset.operational_keys[i].value, sizeof(keyset.operational_keys[i].value));
```
Comment (bzbarsky-apple) [inline src/credentials/GroupDataProviderImpl.cpp:1610]:
```suggestion
        MutableByteSpan key_span(keyset.operational_keys[i].value);
```

File: src/credentials/GroupDataProviderImpl.cpp [hunk 2]
```diff
@@ -1606,12 +1607,11 @@ CHIP_ERROR GroupDataProviderImpl::SetKeySet(chip::FabricIndex fabric_index, cons
...
         ReturnErrorOnFailure(Crypto::DeriveGroupOperationalKey(epoch_key, key_span));
-        ReturnErrorOnFailure(Crypto::DeriveGroupSessionId(ByteSpan(key, sizeof(key)), keyset.operational_keys[i].hash));
+        ReturnErrorOnFailure(Crypto::DeriveGroupSessionId(key_span, keyset.operational_keys[i].hash));
+        VerifyOrReturnError(memcmp(kZeroKey, key_span.data(), sizeof(kZeroKey)), CHIP_ERROR_INTERNAL);
```
Comment (bzbarsky-apple) [inline src/credentials/GroupDataProviderImpl.cpp:1613]: I ... don't understand what this is doing.  `kZeroKey` is an array of `uint8_t *`.  Is it meant to be that?  Are we trying to make sure that `key_span.data()` is not all zeros?

File: src/crypto/CHIPCryptoPAL.h [hunk 3]
```diff
@@ -1355,24 +1355,26 @@ class SymmetricKeyContext
     virtual ~SymmetricKeyContext() = default;
     /**
      * @brief Perform the message encryption as described in 4.7.2. (Security Processing of Outgoing Messages)
-     * @param[in,out] plaintext     Outgoing message payload.
+     * @param[in] plaintext     Outgoing message payload.
      * @param[in] aad           Additional data (message header contents)
      * @param[in] nonce         Nonce (Security Flags | Message Counter | Source Node ID)
-     * @param[out] out_mic      Outgoing Message Integrity Check
+     * @param[out] mic          Outgoing Message Integrity Check
+     * @param[out] ciphertext   Outgoing encrypted payload
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:1362]: It's worth documenting that it's safe to have `plaintext` and `ciphertext` point to the same memory.

And also that ciphertext must be at least as big as plaintext when this function is called.

File: src/crypto/CHIPCryptoPAL.h [hunk 4]
```diff
@@ -1355,24 +1355,26 @@ class SymmetricKeyContext
...
      * @brief Perform the message decryption as described in 4.7.3.(Security Processing of Incoming Messages)
-     * @param[in,out] ciphertext Incoming encrypted payload
-     * @param[in] aad   Additional data (message header contents)
-     * @param[in] nonce Nonce (Security Flags | Message Counter | Source Node ID)
-     * @param[in] mic   Incoming Message Integrity Check
+     * @param[in] ciphertext    Incoming encrypted payload
+     * @param[in] aad           Additional data (message header contents)
+     * @param[in] nonce         Nonce (Security Flags | Message Counter | Source Node ID)
+     * @param[in] mic           Incoming Message Integrity Check
+     * @param[out] plaintext     Incoming message payload
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:1373]: Again, should document that plaintext should be at least as big as ciphertext and is allowed to be the same buffer.

File: src/credentials/GroupDataProviderImpl.cpp (hunk 2 above)
Comment (rcasallas-silabs) [reply src/credentials/GroupDataProviderImpl.cpp:1613]: kZeroKey is intended to be an array of uint8_t. The superfluous pointer is a typo. Fixed now.

File: src/credentials/GroupDataProviderImpl.cpp (hunk 1 above)
Comment (rcasallas-silabs) [reply src/credentials/GroupDataProviderImpl.cpp:1610]: As suggested.

File: src/credentials/GroupDataProviderImpl.cpp [hunk 5]
```diff
@@ -1606,12 +1607,11 @@ CHIP_ERROR GroupDataProviderImpl::SetKeySet(chip::FabricIndex fabric_index, cons
     for (size_t i = 0; i < in_keyset.num_keys_used; ++i)
     {
         ByteSpan epoch_key(in_keyset.epoch_keys[i].key, Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES);
-        uint8_t key[Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
-        MutableByteSpan key_span(key, sizeof(key));
+        MutableByteSpan key_span(keyset.operational_keys[i].value);
         ReturnErrorOnFailure(Crypto::DeriveGroupOperationalKey(epoch_key, key_span));
-        ReturnErrorOnFailure(Crypto::DeriveGroupSessionId(ByteSpan(key, sizeof(key)), keyset.operational_keys[i].hash));
+        ReturnErrorOnFailure(Crypto::DeriveGroupSessionId(key_span, keyset.operational_keys[i].hash));
+        VerifyOrReturnError(memcmp(kZeroKey, key_span.data(), sizeof(kZeroKey)), CHIP_ERROR_INTERNAL);
```
Comment (bzbarsky-apple) [inline src/credentials/GroupDataProviderImpl.cpp:1613]: OK, so now it's clear we are making sure `key_span` is not all zeros.  Then I have the following question:

1) Why are we making sure of that?  Where in the spec is an all-0 operational key disallowed?
2) Should we do this check before we use `key_span` to derive the session id?

Comment (rcasallas-silabs) [reply src/credentials/GroupDataProviderImpl.cpp:1613]: The argument here is that an all-zeroes operation key is a weak key, and must be the result of a faulty key derivation function. This is why the check is perform after the derivation.
I am not sure if a valid epoch key may result in an all-zeroes operational key, but even if such key exist, we may want to reject it, in the grounds that a weak operational key must never be used.

Comment (turon) [reply src/credentials/GroupDataProviderImpl.cpp:1613]: The entire key space is valid.  If we disallow any keys, then we need error conditions on key inputs which creates a surface area for oracle attacks.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by turon — APPROVED** [review]

**Review by turon — APPROVED** [review]
:+1: LGTM

**Pull request merged**